### PR TITLE
PS2 Auto Boot

### PR DESCRIPTION
**Pull request by bbsan2k**
This PR will enable autoboot for FMCB in adding a first stage bootloader that is immediately available after startup.
Note: This won't work for usb debug version, as it waits 2 seconds before actually doing the startup.

### CODE REVIEW

File: src/gui.c [hunk 1]
```diff
@@ -699,7 +729,9 @@ void gui_do_ps2_card_switch(void) {
     input_flush();
 }
 
+
 void gui_task(void) {
+
```
Comment (xyzz) [inline src/gui.c:734]: remove these new empty lines

File: src/ps2/ps2_memory_card.c [hunk 2]
```diff
@@ -48,6 +55,27 @@ uint32_t readptr, writeptr;
 static volatile int mc_exit_request, mc_exit_response, mc_enter_request, mc_enter_response;
 static uint8_t hostkey[9];
 
+static inline void __time_critical_func(read_mc)(uint32_t addr, void *buf, size_t sz) {
+    if (flash_mode) {
+        // Skip first 4 Bytes in Buffer, as they are the prefix
```
Comment (xyzz) [inline src/ps2/ps2_memory_card.c:60]: this comment doesn't seem to correspond to the code

File: src/ps2/ps2_memory_card.c [hunk 3]
```diff
@@ -48,6 +55,27 @@ uint32_t readptr, writeptr;
...
+        ps2_exploit_read(addr, buf, sz);
+        ps2_dirty_unlock();
+    } else {
+        psram_read_dma(addr, buf, sz);
+    }
+}
+
+static inline void __time_critical_func(write_mc)(uint32_t addr, void *buf, size_t sz) {
+    debug_printf("Write at address %d size %d\n", addr, sz);
```
Comment (xyzz) [inline src/ps2/ps2_memory_card.c:69]: remove any debugging output from hot paths e.g. card read/writes

File: src/settings.c [hunk 4]
```diff
@@ -122,3 +123,17 @@ void settings_set_mode(int mode) {
...
+bool settings_get_ps2_autoboot(void) {
+    return (settings.ps2_flags & SETTINGS_FLAGS_AUTOBOOT);
+}
+
+void settings_set_ps2_autoboot(bool autoboot) {
+    if (((settings.ps2_flags & SETTINGS_FLAGS_AUTOBOOT) != 0) != autoboot)     {
+        if (autoboot)
+            settings.ps2_flags |= SETTINGS_FLAGS_AUTOBOOT;
+        else
+            settings.ps2_flags &= ~SETTINGS_FLAGS_AUTOBOOT;
```
Comment (xyzz) [inline src/settings.c:136]:
```
if (autoboot != settings_get_ps2_autoboot())
  settings.ps2_flags ^= SETTINGS_FLAGS_AUTOBOOT;
```

**Review by xyzz — CHANGES_REQUESTED** [review]

File: src/gui.c [hunk 5]
```diff
@@ -448,6 +457,28 @@ static void create_cardswitch_screen(void) {
     lv_label_set_text(g_progress_text, "Read XXX kB/s");
 }
 
+static void create_exploit_screen(void) {
```
[the remark this answers is not in the document]
Comment (bbsan2k) [reply src/gui.c]: Done

**Comment by bbsan2k** [comment]
Issues resolved

**Comment by xyzz** [comment]
So a couple suggestions regarding the usability:

- I don't think there's a need to support multiple cards under BOOT, it's fine to just have a single one; the autoboot will force channel=1 anyway so it's not like you can easily use the other ones; and more importantly...
- there should be a way to switch off from the BOOT card much easier. I think a button press is reasonable (you could hook into the regular card switch logic rather than tracking button clicks specifically I suppose). For example if I'm playing a game and saving to card id=137, slot=4, and then restart my console, it would be a huge pain to navigate back to it after using autoboot. In the current implementation it basically removes the purpose behind saving last-used card.

**Comment by xyzz** [comment]
it needs to work on short press also, it's a bit annoying to long press every boot

**Comment by bbsan2k** [comment]
Here you go!

**Pull request merged**